stor/dx.py

```diff
@@ -144,41 +138,273 @@ def download_objects(self):  # may not need it
+        """
+        if not self.isfile():
+            raise ValueError('can only call download_object on object path')
+        dxpy.download_dxfile(
+            dxid=self.canonical_resource,
+            filename=dest,
+            project=self.canonical_project
+        )
 
     def download(self, dest):
-        """Download a directory."""
-        raise NotImplementedError
+        """Download a directory.
 
-    def upload(self, source):
-        """Upload a list of files and directories to a directory."""
-        raise NotImplementedError
+        Args:
+            dest (Path): The output directory
+
+        Raises:
+            ValueError: When source or dest path is not a directory
+        """
+        if not self.isdir():
+            raise ValueError('can only call download on folder or project paths')
+        dxpy.download_folder(
+            project=self.canonical_project,
+            destdir=dest,
+            folder=self.resource or '/'
+        )
+
+    def upload(self, to_upload):
+        """Upload a list of files and directories to a directory.
+
+        Note that unlike swift, this is not a batch level operation.
+        If some file errors, the files before will remain uploaded
+
+        Args:
+            to_upload (List): A list of file names, directory names, or
+                OBSUploadObject objects to upload.
+
+        Raises:
+            ValueError: When source path is not a directory
+        """
+        dx_upload_objects = [
+            name for name in to_upload
+            if isinstance(name, OBSUploadObject)
+        ]
+        all_files_to_upload = utils.walk_files_and_dirs([
+            name for name in to_upload
+            if not isinstance(name, OBSUploadObject)
+        ])
+        dx_upload_objects.extend([
+            OBSUploadObject(f,
+                            object_name=(self.resource or Path(''))
+                            / utils.file_name_to_object_name(f))
+            for f in all_files_to_upload
+        ])
+        for upload_obj in dx_upload_objects:
+            upload_obj.object_name = Path(upload_obj.object_name)
+            upload_obj.source = Path(upload_obj.source)
+            dest_file = Path('{drive}{project}:{path}'.format(
+                drive=self.drive, project=self.canonical_project,
+                path=upload_obj.object_name))
+
+            dest_exists = dest_file.isdir()
+            if dest_exists:
+                new_dest = dest_file / upload_obj.source.name
+                if new_dest.isfile():
+                    raise ValueError('Dest path already exists. Do not duplicate!')
+                folder_dest = dest_file
+                file_name = upload_obj.source.name
+            else:
+                folder_dest = dest_file.parent
+                file_name = upload_obj.object_name.name
+            print(upload_obj.source, upload_obj.object_name.name, folder_dest.resource)
+            dxpy.upload_local_file(
```

Comment:
can we/ should we parallelize this at all? Perhaps this is a worthy place to put in a configuration option and then use concurrent.futures to manage a thread or process pool?

Reply:
We could/ we should. Let's do it as optimization in stor-dx v2? I can put up a JIRA ticket for it.

Reply:
Closing this. This function is called per object, and parallelism won't help here. Note that for downloads, dxpy already works in parallel.